daemon/algod/api/server/v2/handlers.go

```diff
@@ -291,18 +301,21 @@ func (v2 *Handlers) GenerateParticipationKeys(ctx echo.Context, address string,
 		err := fmt.Errorf("participation key generation already in progress")
 		return badRequest(ctx, err, err.Error(), v2.Log)
 	}
+	// Generate the keys on the main thread to block response
+	// KeysBuilder already handles coroutines, this response feedback
+	// is necessary for end users
+	part, err := v2.generateKeyHandler(address, params)
+	if err != nil {
+		return badRequest(ctx, err, err.Error(), v2.Log)
+	}
 
 	// Semaphore was acquired, generate the key.
 	go func() {
 		defer v2.KeygenLimiter.Release(1)
-		err := v2.generateKeyHandler(address, params)
-		if err != nil {
-			v2.Log.Warnf("Error generating participation keys: %v", err)
-		}
 	}()
```

Comment:
This starts a (nearly empty) go routine which immediately releases the `KeygenLimiter` semaphore. Therefore, this allows any number of keygens to happen concurrently.  I doubt that's what we want, so the `defer` line should be moved outside of the go routine, which should go away. I would put it before line 307.

Since generating part keys can take a very long time, I'm not sure what will happen to the TCP connection that's being held open the entire time. I expect timeouts will occur in many situations.

Reply:
This was the only was I could get the lock while testing. I will take another look at it.

Reply:
Somewhat off-topic. It's unclear to me why a `semaphore.Weighted` was used. It seems to be used like a simple `sync.WaitGroup`.  You don't have to fix that, I suppose.